File: chapter-02-auth/python/graph_auth_example.py
"""
graph_auth_example.py
Chapter 02: Authentication

Microsoft Graph Client Credentials Authentication Example
Demonstrates authenticating to Microsoft Graph using client credentials flow

Required environment variables:
    - TENANT_ID
    - CLIENT_ID
    - CLIENT_SECRET
"""

import os
import sys
from typing import Dict, Optional
import requests
from urllib.parse import urlencode


class GraphAuthConfig:
    """Configuration for Microsoft Graph authentication"""

    def __init__(self):
        self.tenant_id = os.environ.get('TENANT_ID')
        self.client_id = os.environ.get('CLIENT_ID')
        self.client_secret = os.environ.get('CLIENT_SECRET')

        self._validate_config()

    def _validate_config(self) -> None:
        """Validates that all required environment variables are present"""
        required = ['TENANT_ID', 'CLIENT_ID', 'CLIENT_SECRET']
        missing = [key for key in required if not os.environ.get(key)]

        if missing:
            raise ValueError(f"Missing required environment variables: {', '.join(missing)}")


class GraphAuthExample:
    """Microsoft Graph Authentication handler using client credentials"""

    def __init__(self):
        self.config = GraphAuthConfig()
        self.token_endpoint = (
            f"https://login.microsoftonline.com/{self.config.tenant_id}/oauth2/v2.0/token"
        )
        self.scope = "https://graph.microsoft.com/.default"

    def get_access_token(self) -> str:
        """
        Acquires an access token using client credentials flow

        Returns:
            str: Access token for Microsoft Graph API
        """
        try:
            request_data = {
                'client_id': self.config.client_id,
                'client_secret': self.config.client_secret,
                'scope': self.scope,
                'grant_type': 'client_credentials'
            }

            response = requests.post(
                self.token_endpoint,
                data=request_data,
                headers={'Content-Type': 'application/x-www-form-urlencoded'}
            )
            response.raise_for_status()

            print("Successfully authenticated to Microsoft Graph")
            return response.json()['access_token']

        except requests.exceptions.RequestException as e:
            print(f"Authentication failed: {e}")
            raise Exception("Failed to acquire access token") from e

    def get_authenticated_headers(self) -> Dict[str, str]:
        """
        Creates authenticated API client headers

        Returns:
            Dict[str, str]: Headers dictionary with authorization
        """
        token = self.get_access_token()
        return {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }

    def test_connection(self) -> Optional[Dict]:
        """
        Tests the connection by retrieving organization details

        Returns:
            Optional[Dict]: Organization information or None if failed
        """
        try:
            headers = self.get_authenticated_headers()
            response = requests.get(
                'https://graph.microsoft.com/v1.0/organization',
                headers=headers
            )
            response.raise_for_status()

            org = response.json()['value'][0]
            print(f"Connected to tenant: {org['displayName']}")
            return response.json()

        except requests.exceptions.RequestException as e:
            print(f"Connection test failed: {e}")
            raise


def main():
    """Main execution function"""
    try:
        print("=== Microsoft Graph Authentication Example ===")

        auth = GraphAuthExample()
        auth.test_connection()

        print("\nAuthentication completed successfully!")

    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: chapter-02-auth/javascript/GraphAuthExample.js
/**
 * GraphAuthExample.js
 * Chapter 02: Authentication
 *
 * Microsoft Graph Client Credentials Authentication Example
 * Demonstrates authenticating to Microsoft Graph using client credentials flow
 *
 * Required environment variables:
 * - TENANT_ID
 * - CLIENT_ID
 * - CLIENT_SECRET
 */

const axios = require('axios');
const qs = require('querystring');

/**
 * Authentication configuration class
 */
class GraphAuthConfig {
    constructor() {
        this.tenantId = process.env.TENANT_ID;
        this.clientId = process.env.CLIENT_ID;
        this.clientSecret = process.env.CLIENT_SECRET;

        this.validateConfig();
    }

    /**
     * Validates that all required environment variables are present
     */
    validateConfig() {
        const required = ['TENANT_ID', 'CLIENT_ID', 'CLIENT_SECRET'];
        const missing = required.filter(key => !process.env[key]);

        if (missing.length > 0) {
            throw new Error(`Missing required environment variables: ${missing.join(', ')}`);
        }
    }
}

/**
 * Microsoft Graph Authentication handler
 */
class GraphAuthExample {
    constructor() {
        this.config = new GraphAuthConfig();
        this.tokenEndpoint = `https://login.microsoftonline.com/${this.config.tenantId}/oauth2/v2.0/token`;
        this.scope = 'https://graph.microsoft.com/.default';
    }

    /**
     * Acquires an access token using client credentials flow
     * @returns {Promise<string>} Access token
     */
    async getAccessToken() {
        try {
            const requestBody = {
                client_id: this.config.clientId,
                client_secret: this.config.clientSecret,
                scope: this.scope,
                grant_type: 'client_credentials'
            };

            const response = await axios.post(
                this.tokenEndpoint,
                qs.stringify(requestBody),
                {
                    headers: {
                        'Content-Type': 'application/x-www-form-urlencoded'
                    }
                }
            );

            console.log('Successfully authenticated to Microsoft Graph');
            return response.data.access_token;
        } catch (error) {
            console.error('Authentication failed:', error.response?.data?.error_description || error.message);
            throw new Error('Failed to acquire access token');
        }
    }

    /**
     * Creates authenticated API client headers
     * @returns {Promise<Object>} Headers object with authorization
     */
    async getAuthenticatedHeaders() {
        const token = await this.getAccessToken();
        return {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json'
        };
    }

    /**
     * Tests the connection by retrieving organization details
     */
    async testConnection() {
        try {
            const headers = await this.getAuthenticatedHeaders();
            const response = await axios.get('https://graph.microsoft.com/v1.0/organization', { headers });

            const org = response.data.value[0];
            console.log(`Connected to tenant: ${org.displayName}`);
            return response.data;
        } catch (error) {
            console.error('Connection test failed:', error.response?.data?.error?.message || error.message);
            throw error;
        }
    }
}

/**
 * Main execution function
 */
async function main() {
    try {
        console.log('=== Microsoft Graph Authentication Example ===');

        const auth = new GraphAuthExample();
        await auth.testConnection();

        console.log('\nAuthentication completed successfully!');
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

// Export for module usage
module.exports = { GraphAuthExample, GraphAuthConfig };

// Run if executed directly
if (require.main === module) {
    main();
}
